src/autoscaler/db/sqldb/policy_sqldb.go

```diff
@@ -145,3 +146,41 @@ func (pdb *PolicySQLDB) DeletePolicy(appId string) error {
 	}
 	return err
 }
+
+func (pdb *PolicySQLDB) GetCustomMetricsCreds(appId string) (string, string, error) {
+	var password string
+	var username string
+	query := "SELECT username,password from credentials WHERE id = $1"
+	err := pdb.sqldb.QueryRow(query, appId).Scan(&username, &password)
+
+	if err != nil {
+		pdb.logger.Error("get-custom-metrics-creds-from-credentials-table", err, lager.Data{"query": query})
+		return "", "", err
+	}
+	return username, password, nil
+}
+
+func (pdb *PolicySQLDB) ValidateCustomMetricTypes(appGUID string, metricsConsumer *models.MetricsConsumer) (bool, error) {
+	scalingPolicy, err := pdb.GetAppPolicy(appGUID)
+	if err != nil {
+		pdb.logger.Error("errorr-getting-policy", err, lager.Data{"appId": appGUID})
+		return false, errors.New("not able to get policy details")
+	}
+	if err == nil && scalingPolicy == nil {
+		pdb.logger.Debug("no-policy-found", lager.Data{"appId": appGUID})
+		return false, errors.New("no policy found")
+	}
+	allowedMetricTypeSet := make(map[string]struct{}, len(scalingPolicy.ScalingRules))
+	for _, metrictype := range scalingPolicy.ScalingRules {
+		allowedMetricTypeSet[metrictype.MetricType] = struct{}{}
+	}
+	pdb.logger.Debug("allowed-metrics-types", lager.Data{"metrics": allowedMetricTypeSet})
+	for _, metric := range metricsConsumer.CustomMetrics {
```

Comment:
Also, I guess we need to cache the allowed metric type for an application to avoid query & parsing policy on each request

Reply:
@boyang9527 , @cdlliuy  IMHO it's not a good idea to allow user to emit any custom metrics through autoscaler without defining the metrics in the policy.

Could you please explain the use case we are targeting here ?

@pradyutsarma  : Your thoughts ?

Reply:
the user may emit the custom metrics without a rule defined. This will give the user the possibility to run performance test, watch how the metrics go with the workload and identify the right upper and lower threshold.